File: python/infinicore/ops/addbmm.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def addbmm(input, batch1, batch2, *, beta=1.0, alpha=1.0, out=None):
    # 1. Out-of-place 模式 (如果没有指定 out)
    if out is None:
        return Tensor(
            _infinicore.addbmm(
                input._underlying, batch1._underlying, batch2._underlying, beta, alpha
            )
        )

    # 2. In-place 模式 (指定了 out)
    _infinicore.addbmm_(
        out._underlying,
        input._underlying,
        batch1._underlying,
        batch2._underlying,
        beta,
        alpha,
    )

    return out

File: ops/src/infinicore/ops/addbmm/addbmm.cc
#include "infinicore/ops/addbmm.hpp"
#include "../../utils.hpp"
#include "infinicore/ops/addbmm.hpp"

namespace infinicore::op {

// 1. 初始化 Dispatcher
common::OpDispatcher<Addbmm::schema> &Addbmm::dispatcher() {
    static common::OpDispatcher<Addbmm::schema> dispatcher_;
    return dispatcher_;
};

void Addbmm::execute(Tensor output, Tensor input, Tensor batch1, Tensor batch2, float beta, float alpha) {

    // 切换上下文
    infinicore::context::setDevice(output->device());

    // 分发计算
    dispatcher().lookup(output->device().getType())(output, input, batch1, batch2, beta, alpha);
}

Tensor addbmm(Tensor input, Tensor batch1, Tensor batch2, float beta, float alpha) {
    auto output = Tensor::empty(input->shape(), input->dtype(), input->device());
    Addbmm::execute(output, input, batch1, batch2, beta, alpha);
    return output;
}

void addbmm_(Tensor output, Tensor input, Tensor batch1, Tensor batch2, float beta, float alpha) {
    Addbmm::execute(output, input, batch1, batch2, beta, alpha);
}

} // namespace infinicore::op
